src/path.c

```diff
@@ -264,6 +264,66 @@ ly_path_check_predicate(const struct ly_ctx *ctx, const struct lysc_node *cur_no
     return LY_EVALID;
 }
 
+/**
+ * @brief Parse deref XPath function and perform all additional checks.
+ *
+ * @param[in] ctx libyang context.
+ * @param[in] ctx_node Optional context node, used for logging.
+ * @param[in] exp Parsed path.
+ * @param[in,out] tok_idx Index in @p exp, is adjusted.
+ * @return LY_ERR value.
+ */
+static LY_ERR
+ly_path_parse_deref(const struct ly_ctx *ctx, const struct lysc_node *ctx_node, const struct lyxp_expr *exp, uint32_t *tok_idx)
+{
+    LY_ERR ret = LY_SUCCESS;
+    size_t arg_len;
+    uint32_t begin_token, end_token;
+    struct lyxp_expr *arg_expr = NULL;
+
+    /* mandatory FunctionName */
+    LY_CHECK_ERR_GOTO(lyxp_next_token(ctx, exp, tok_idx, LYXP_TOKEN_FUNCNAME), ret = LY_EVALID, error);
+    if (exp->used <= *tok_idx + 1) {
+        if (ctx) {
+            LOGVAL(ctx, LY_VCODE_XP_EOF);
+        }
+        return LY_EINCOMPLETE;
+    }
+
+    if (!strncmp(&exp->expr[exp->tok_pos[*tok_idx]], "deref", 5)) {
+        LOGVAL(ctx, LYVE_XPATH, "Unexpected XPath function \"%.*s\" in path, expected \"deref(...)\"", exp->tok_len[*tok_idx], exp->tok_pos[*tok_idx]);
+        ret = LY_EVALID;
+        goto error;
+    }
+
+    /* mandatory '(' */
+    LY_CHECK_ERR_GOTO(lyxp_next_token(ctx, exp, tok_idx, LYXP_TOKEN_PAR1), ret = LY_EVALID, error);
+    begin_token = *tok_idx;
+
+    /* count tokens till ')' */
+    while (lyxp_check_token(NULL, exp, *tok_idx, LYXP_TOKEN_PAR2) && lyxp_check_token(NULL, exp, *tok_idx, LYXP_TOKEN_FUNCNAME)) {
+        (*tok_idx)++;
+    }
+
+    /* embedded function inside function ')' */
+    if (!lyxp_check_token(NULL, exp, *tok_idx, LYXP_TOKEN_FUNCNAME)) {
+        LOGVAL(ctx, LYVE_XPATH, "Embedded function XPath function inside deref function within the path is not allowed");
+        ret = LY_EVALID;
+        goto error;
+    }
+
+    /* mandatory ')' */
+    LY_CHECK_ERR_GOTO(lyxp_next_token(ctx, exp, tok_idx, LYXP_TOKEN_PAR2), ret = LY_EVALID, error);
+    end_token = *tok_idx - 1;
+
+    /* verify the path of deref argument */
+    arg_len = exp->tok_pos[end_token] - exp->tok_pos[begin_token];
+    LY_CHECK_ERR_GOTO(ly_path_parse(ctx, ctx_node, &exp->expr[exp->tok_pos[begin_token]], arg_len, 1, LY_PATH_BEGIN_EITHER, LY_PATH_PREFIX_OPTIONAL, LY_PATH_PRED_LEAFREF, &arg_expr), ret = LY_EVALID, error);
+    lyxp_expr_free(ctx, arg_expr);
+error:
```

Comment:
Just some customs we have, if the label code is always performed, we call it `cleanup` and there should always be an empty line before the label (there are a few instances of this being broken).

Reply:
Done